scripts/layerops.py

```diff
@@ -0,0 +1,721 @@
+                temp[entry] = data[entry]._get('techniques')
+            self.mode = 'dict'
+            self._merge_to_template(key=list(data.keys())[0])
+        elif isinstance(data, list):
+            temp = []
+            for entry in data:
+                temp.append(entry._get('techniques'))
+            self.mode = 'list'
+            self._merge_to_template()
+        else:
+            raise InvalidFormat
+        self.da = temp
+        self.corpus = self._build_template(temp)
+
+        # Default values for fields if they aren't supplied
+        self.default_values = {
+            "comment": "",
+            "enabled": True,
+            "color": "#ffffff",
+            "score": 1,
+            "metadata": []
+        }
+        self.processed = None
+
+    """
+        Processes the ingested data using User provided Lambdas
+        :param score: lambda to calculate score
+        :param comment: lambda to generate comments
+        :param enabled: lambda to determine enabled status
+        :param metadata: lambda to generate metadata
+        :param defaults: dictionary containing desired default values for missing data element values
+        :param out_Name: new name to apply to the resulting layer
+        :param out_Desc: new description to apply to the resulting layer
+    """
+    def process(self, score=None, comment=None, enabled=None, colors=None, metadata=None, defaults=None,
+                out_Name=None, out_Desc=None):
+        composite = copy.deepcopy(self.corpus)
+        if score != None:
+            for entry in composite:
+                entry['score'] = self._applyOperation(self.da, entry, 'score', score, defaults)
+
+        if comment != None:
+            for entry in composite:
+                entry['comment'] = self._applyOperation(self.da, entry, 'comment', comment, defaults)
+
+        if enabled != None:
+            for entry in composite:
+                entry['enabled'] = self._applyOperation(self.da, entry, 'enabled', enabled, defaults)
+
+        if colors != None:
+            for entry in composite:
+                entry['color'] = self._applyOperation(self.da, entry, 'color', colors, defaults)
+
+        if metadata != None:
+            for entry in composite:
+                entry['metadata'] = self._applyOperation(self.da, entry, 'metadata', metadata, defaults)
+
+        processed = copy.deepcopy(self.out)
+        processed['techniques'] = composite
+        if out_Name != None:
+            processed['name'] = out_Name
+        if out_Desc != None:
+            processed['description'] = out_Desc
+        self.processed = Layer(processed)
+
+    """
+        Retrieve processed results
+        :returns: A Layer object representing the processing changes
+    """
+    def get_processed(self):
+        if self.processed is not None:
+            return self.processed
+        else:
+            print('Please call the Process method to combine loaded layers.')
```

Comment:
I'm not sure I understand your approach here. Why does the user need a separate step to get the output? Why not just have process return the processed layer directly? In fact, why is there even a class based implementation in the first place?

---

The original API outlined in #2 is pretty different than what you've implemented. There is definitely merit to your class based approach, but the way it's done right now seems to require a lot of unnecessary steps on the part of the user. The API in #2 is designed to be very simple. There's no need for the user to instantiate a class, call multiple functions, etc — it's just a single function where the user passes the layers and then the lambda functions for the various allowed combination operations. Layerops doesn't _need_ to have a class according to that design, and therefore the user doesn't need to do the extra work of instantiating a Layerops instance.

---

Now, the class based approach you've taken could've been done in a really nifty way. A user could define a Layerops instance, in the constructor define the set of operations to be done in the combination. Then, in the "combine" function, the user would pass the layers to be combined as the parameter, and the returned value would be the defined set of operations applied to the parameter layers. 

With this implementation, the user could define a "average scores" Layerops instance or what have you, and easily apply it to various sets of layers without having to redefine the operations required every time. 

However, by having the layers as the parameter to the constructor, and the operations the parameter to the process function, as far as I can tell the user gets no benefit from having a Layerops instance over the simpler API defined in #2.

Maybe I'm missing something about the class based approach?

Reply:
Thinking more about this, I actually quite like the "constructor defines the operations to be performed on layers, and the function call performs the operations on the parameter layers" design. It'd work well for #1 as well (Define the renderer properties once, run that same instance on a bunch of similar layers).

Reply:
I think I've got a viable option for this.